File: src/main/java/com/cxxcxx/zinecraft/api/world/structure/FixedOriginStructurePlacement.java
package com.cxxcxx.zinecraft.api.world.structure;

import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.core.Vec3i;
import net.minecraft.world.level.chunk.ChunkGeneratorStructureState;
import net.minecraft.world.level.levelgen.structure.placement.StructurePlacement;
import net.minecraft.world.level.levelgen.structure.placement.StructurePlacementType;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.Optional;
import java.util.function.Supplier;

public final class FixedOriginStructurePlacement extends StructurePlacement {
  @NotNull
  public static final FixedOriginStructurePlacement.Access ACCESS = new FixedOriginStructurePlacement.Access();
  @NotNull
  private static final MapCodec<FixedOriginStructurePlacement> CODEC;
  private static Supplier<StructurePlacementType<FixedOriginStructurePlacement>> type;

  static {
    CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
        com.mojang.serialization.Codec.INT.optionalFieldOf("chunk_x", -1)
            .forGetter(placement -> placement.chunkX),
        com.mojang.serialization.Codec.INT.optionalFieldOf("chunk_z", -1)
            .forGetter(placement -> placement.chunkZ)
    ).apply(instance, FixedOriginStructurePlacement::new));
  }

  private final int chunkX;
  private final int chunkZ;

  private FixedOriginStructurePlacement(int chunkX, int chunkZ) {
    super(new Vec3i(16, 0, 16), FrequencyReductionMethod.DEFAULT, 1.0F, 0, Optional.empty());
    this.chunkX = chunkX;
    this.chunkZ = chunkZ;
  }

  protected boolean isPlacementChunk(@NotNull ChunkGeneratorStructureState structureState, int chunkX, int chunkZ) {
    return chunkX == this.chunkX && chunkZ == this.chunkZ;
  }

  @NotNull
  public StructurePlacementType<?> type() {
    return Objects.requireNonNull(type, "Fixed-origin structure placement is not registered").get();
  }

  public static final class Access {
    private Access() {
    }

    @NotNull
    public final MapCodec<FixedOriginStructurePlacement> getCODEC() {
      return FixedOriginStructurePlacement.CODEC;
    }

    public void bind(Supplier<StructurePlacementType<FixedOriginStructurePlacement>> type) {
      if (FixedOriginStructurePlacement.type == null) {
        FixedOriginStructurePlacement.type = Objects.requireNonNull(type, "type");
      }
    }

    @NotNull
    public final FixedOriginStructurePlacement create() {
      return create(-1, -1);
    }

    /**
     * @param chunkX 唯一生成区块 X
     * @param chunkZ 唯一生成区块 Z
     * @return 固定在指定区块的结构放置器
     */
    @NotNull
    public final FixedOriginStructurePlacement create(int chunkX, int chunkZ) {
      if (FixedOriginStructurePlacement.type == null) {
        throw new IllegalStateException("固定坐标结构放置器尚未注册");
      } else {
        return new FixedOriginStructurePlacement(chunkX, chunkZ);
      }
    }
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/api/world/structure/FixedOriginStructurePlacement.kt
package com.cxxcxx.zinecraft.api.world.structure

import com.cxxcxx.zinecraft.api.registry.ModRegistrar
import com.mojang.serialization.MapCodec
import net.minecraft.core.Vec3i
import net.minecraft.world.level.chunk.ChunkGeneratorStructureState
import net.minecraft.world.level.levelgen.structure.placement.StructurePlacement
import net.minecraft.world.level.levelgen.structure.placement.StructurePlacementType
import java.util.*

/**
 * 将结构唯一候选区块固定在 `(-1, -1)`。
 *
 * 33×33 模板由该区块的最小坐标 `(-16, -16)` 展开后，几何中心恰好落在世界方块坐标 `(0, 0)`；定位偏移也指向原点。
 */
class FixedOriginStructurePlacement private constructor() : StructurePlacement(
  Vec3i(16, 0, 16),
  FrequencyReductionMethod.DEFAULT,
  1.0f,
  0,
  Optional.empty()
) {
  override fun isPlacementChunk(
    structureState: ChunkGeneratorStructureState,
    chunkX: Int,
    chunkZ: Int
  ): Boolean = chunkX == ORIGIN_CHUNK && chunkZ == ORIGIN_CHUNK

  override fun type(): StructurePlacementType<*> = TYPE

  companion object {
    private const val ORIGIN_CHUNK = -1

    val CODEC: MapCodec<FixedOriginStructurePlacement> = MapCodec.unit(::FixedOriginStructurePlacement)

    private lateinit var TYPE: StructurePlacementType<FixedOriginStructurePlacement>

    internal fun register(registrar: ModRegistrar) {
      if (!::TYPE.isInitialized) {
        TYPE = registrar.structurePlacement("fixed_origin", CODEC)
      }
    }

    internal fun create(): FixedOriginStructurePlacement {
      check(::TYPE.isInitialized) { "固定原点结构放置器尚未注册" }
      return FixedOriginStructurePlacement()
    }
  }
}
